src/rust/cryptography-x509-validation/src/policy/extension.rs

```diff
@@ -0,0 +1,366 @@
+            (ExtensionValidator::NotPresent, Some(_)) => Err(PolicyError::Other(
+                "EE certificate contains prohibited extension",
+            )),
+            // Extension MUST be present but is not; NOT OK.
+            (ExtensionValidator::Present { .. }, None) => Err(PolicyError::Other(
+                "EE certificate is missing required extension",
+            )),
+            // Extension MUST be present and is; check it.
+            (
+                ExtensionValidator::Present {
+                    criticality,
+                    validator,
+                },
+                Some(extn),
+            ) => {
+                if !criticality.permits(extn.critical) {
+                    return Err(PolicyError::Other(
+                        "EE certificate extension has incorrect criticality",
+                    ));
+                }
+
+                // If a custom validator is supplied, apply it.
+                validator.map_or(Ok(()), |v| v(policy, cert, &extn))
+            }
+            // Extension MAY be present.
+            (
+                ExtensionValidator::MaybePresent {
+                    criticality,
+                    validator,
+                },
+                extn,
+            ) => {
+                // If the extension is present, apply our criticality check.
+                if extn
+                    .as_ref()
+                    .map_or(false, |extn| !criticality.permits(extn.critical))
+                {
+                    return Err(PolicyError::Other(
+                        "EE certificate extension has incorrect criticality",
+                    ));
+                }
+
+                // If a custom validator is supplied, apply it.
+                validator.map_or(Ok(()), |v| v(policy, cert, extn.as_ref()))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Criticality, ExtensionPolicy};
+    use crate::ops::tests::{cert, v1_cert_pem, NullOps};
+    use crate::ops::CryptoOps;
+    use crate::policy::{Policy, PolicyError};
+    use asn1::{ObjectIdentifier, SimpleAsn1Writable};
+    use cryptography_x509::certificate::Certificate;
+    use cryptography_x509::extensions::{
+        BasicConstraints, Extension, Extensions, PolicyConstraints,
+    };
+    use cryptography_x509::oid::BASIC_CONSTRAINTS_OID;
+
+    #[test]
+    fn test_criticality_critical() {
+        let criticality = Criticality::Critical;
+        assert!(criticality.permits(true));
+        assert!(!criticality.permits(false));
+    }
+
+    #[test]
+    fn test_criticality_agnostic() {
+        let criticality = Criticality::Agnostic;
+        assert!(criticality.permits(true));
+        assert!(criticality.permits(false));
+    }
+
+    #[test]
+    fn test_criticality_non_critical() {
+        let criticality = Criticality::NonCritical;
+        assert!(!criticality.permits(true));
+        assert!(criticality.permits(false));
+    }
```

Comment:
Extreme nitpick, but we could save a few LoC here by collapsing these into a single `test_criticality_variants` or similar.

Reply:
[33e1cf5](https://github.com/pyca/cryptography/pull/9781/commits/33e1cf56b68d874adf70bd0159c861ac3d9f4514)